### PR TITLE
torchdynamo: add convolution add(relu) inplace fusion kernel

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88794
* #88414
* #88412
* #88403
* __->__ #88048

This PR is about add convolution add(relu) inplace fusion kernel which  works for **other.add_(conv)**.

cc @VitalyFedyunin @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 1]
```diff
@@ -416,32 +444,53 @@ Tensor mkldnn_convolution_pointwise_binary(
...
+    c10::optional<c10::string_view> unary_attr,
+    torch::List<c10::optional<at::Scalar>> unary_scalars,
+    c10::optional<c10::string_view> unary_algorithm) {
+  // other_t += convolution(...), other_t = unary(other_t)
   TORCH_CHECK(
       input_t.ndimension() == 4 || input_t.ndimension() == 5,
       "mkldnn_convolution_add_: currently only support 2d and 3d")
+  TORCH_CHECK(
+      binary_attr == "add",
+      "mkldnn_convolution_pointwise_binary_: currently only add inplace fusion")
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp:487]:
```suggestion
      "mkldnn_convolution_pointwise_binary_: only support add binary op fusion")
```

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 2]
```diff
@@ -416,32 +444,53 @@ Tensor mkldnn_convolution_pointwise_binary(
...
+  TORCH_CHECK(
+      !alpha.has_value() || alpha.value().to<float>() == 1.0,
+      "mkldnn_convolution_pointwise_binary: the alpha value should be none or 1.0");
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp:490]:
```suggestion
      "mkldnn_convolution_pointwise_binary: the alpha value for the binary op should be none (meaning 1.0) or 1.0");
```

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 3]
```diff
@@ -416,32 +444,53 @@ Tensor mkldnn_convolution_pointwise_binary(
...
+  TORCH_CHECK(
+      !unary_attr.has_value() || unary_attr.value() == "relu",
+      "mkldnn_convolution_pointwise_binary: the unary_attr value should be none or relu");
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp:493]:
```suggestion
      "mkldnn_convolution_pointwise_binary: only support none or relu unary op fusion after binary op");
```

File: aten/src/ATen/native/mkldnn/Utils.cpp [hunk 4]
```diff
@@ -140,6 +140,13 @@ const std::map<c10::string_view, AttrFunction>& fx_fusion_attr_map() {
   return fusion_attr_map;
 };
 
+const std::map<c10::string_view, ideep::algorithm>& fusion_unary_alg_map() {
+  static const std::map<c10::string_view, ideep::algorithm> fusion_attr_map{
+      {"relu", {ideep::algorithm::eltwise_relu}},
+  };
+  return fusion_attr_map;
+};
+
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Utils.cpp:149]: Why can't we reuse `fx_fusion_attr_map`? Also, should `fx_fusion_attr_map` be renamed to `fusion_unary_attr_map`?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 5]
```diff
@@ -322,10 +322,17 @@ Tensor mkldnn_convolution_pointwise_binary(
     IntArrayRef stride,
     IntArrayRef dilation,
     int64_t groups,
-    c10::string_view attr) {
+    c10::string_view binary_attr,
+    c10::optional<at::Scalar> alpha,
+    c10::optional<c10::string_view> unary_attr,
+    torch::List<c10::optional<at::Scalar>> unary_scalars,
+    c10::optional<c10::string_view> unary_algorithm) {
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Conv.cpp:338]: some explanations are added.

File: aten/src/ATen/native/mkldnn/Utils.cpp (hunk 4 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Utils.cpp:149]: if we reuse **fx_fusion_attr_map**, we need to query the eltwise alg and its paraemters to re-create a new op attr, so I think we can have an alg map as binary and directly query the eltwise alg. I have changed the  **fx_fusion_attr_map** to **fusion_unary_attr_map**.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: test/test_mkldnn_fusion.py [hunk 6]
```diff
@@ -282,13 +282,53 @@ def forward(self, x, other):
...
+                            unary_attr = "relu"
                         attr = pointwise_name
                         fused = torch.ops.mkldnn._convolution_pointwise(
                             x, other, mod.conv.weight, mod.conv.bias, mod.conv.padding, mod.conv.stride, mod.conv.dilation,
-                            mod.conv.groups, attr
+                            mod.conv.groups, attr, None, unary_attr, [], None
                         )
                     self.assertEqual(ref, fused)
 
+    def test_conv_add_relu_inplace(self):
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply test/test_mkldnn_fusion.py]: yes, I changed it by add this test case into test_conv_binary_fusion_ops.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**